### PR TITLE
Use Fastly for cache layer

### PR DESCRIPTION
**Pull request by allenan**
- remove internal stats cache
- add fastly header

### CODE REVIEW

**Comment by vihu** [comment]
You might want to stop and remove the `Cachex` child spec from `application.ex` as well...

File: config/prod.exs [hunk 1]
```diff
@@ -28,6 +28,8 @@ config :blockchain_api,
   google_maps_secret: System.get_env("GOOGLE_MAPS_API_KEY"),
   onesignal_rest_api_key: System.get_env("ONESIGNAL_API_KEY"),
   onesignal_app_id: System.get_env("ONESIGNAL_APP_ID"),
+  fastly_api_key: System.get_env("FASTLY_API_KEY"),
```
Comment (vihu) [inline config/prod.exs]: We should may be put these in `dev.exs` as well for alamo support?

File: lib/blockchain_api/purger.ex [hunk 2]
```diff
@@ -0,0 +1,15 @@
+defmodule BlockchainAPI.Purger do
```
Comment (vihu) [inline lib/blockchain_api/purger.ex]: What do you thing about renaming this module to `BlockchainAPI.Fastly` or `BlockchainAPI.FastlyService` and have the functions be just `purge` and `post`?

File: lib/blockchain_api_web/controllers/hotspot_controller.ex [hunk 3]
```diff
@@ -9,8 +9,11 @@ defmodule BlockchainAPIWeb.HotspotController do
   def index(conn, params) do
     hotspots = Query.Hotspot.list(params)
 
-    render(
-      conn,
+    conn
+    |> put_resp_header("surrogate-key", "block")
```
Comment (vihu) [inline lib/blockchain_api_web/controllers/hotspot_controller.ex]: There seems to be a lot of duplication with the fastly resp headers? Maybe have some sort of fastly header util? which can put headers based on keys and max-age and use that?

That being said it's nice to be verbose for this, I'm not sure, your call.

Comment (allenan) [reply lib/blockchain_api_web/controllers/hotspot_controller.ex]: yeah i was thinking about making some kind of fancy plug, especially for the `?before=x&limit=x` requests. but was just being verbose for now until a pattern emerges

File: lib/blockchain_api/purger.ex (hunk 2 above)
Comment (allenan) [reply lib/blockchain_api/purger.ex]: I _think_ we'd want to not specify the underlying service in the module name. In theory we could switch to a different cache service and the code calling into this module wouldn't need to change. But if we have some header utils, we could broaden this to something like `CacheService` or something like that.

File: config/prod.exs (hunk 1 above)
Comment (allenan) [reply config/prod.exs]: oh hmm does alamo run in dev environment? that seems a bit odd, but yeah I can add it to dev as well. we just wouldn't set those env vars when actually running in dev, but the module's designed to handle that case

Comment (vihu) [reply config/prod.exs]: Yeah, had to separate it out to not do notifications from alamo.

**Comment by vihu** [comment]
Since we no longer use `Cachex`, probably should remove it from `mix.exs` and bump `mix.lock` accordingly. Other than that looks good :+1:

File: lib/blockchain_api/cache/cache_service.ex [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    |> put_resp_header("surrogate-control", "max-age=#{ttl}")
+    |> put_resp_header("cache-control", "max-age=#{ttl}")
+  end
+
+  defp get_ttl_value(ttl) do
+    case ttl do
+      :never -> 0
+      :short -> 300   # 5 minutes
+      :medium -> 600  # 10 minutes
+      :long -> 86_400 # 1 day
```
Comment (vihu) [inline lib/blockchain_api/cache/cache_service.ex]: I'm assuming it's impossible to get any other `ttl` value?

**Pull request merged**